irohad/ametsuchi/impl/postgres_query_executor.cpp

```diff
@@ -342,6 +342,119 @@ namespace iroha {
       return make_error_response(error);
     }
 
+    template <typename Query, typename QueryApplier, typename... Permissions>
+    QueryExecutorResult PostgresQueryExecutorVisitor::executeTransactionsQuery(
+        const Query &q,
+        const std::string &related_txs,
+        QueryApplier applier,
+        Permissions... perms) {
+      using QueryTuple = QueryType<shared_model::interface::types::HeightType,
+                                   uint64_t,
+                                   uint64_t>;
+      using PermissionTuple = boost::tuple<int>;
+      const auto &pagination_info = q.paginationMeta();
+      auto first_hash = pagination_info.firstTxHash();
+      // retrieve one extra transaction to populate next_hash
+      auto query_size = pagination_info.pageSize() + 1u;
+
+      auto base = boost::format(R"(WITH has_perms AS (%s),
+      my_txs AS (%s),
+      first_hash AS (%s),
+      previous_txes AS (
+        SELECT position_by_hash.height, position_by_hash.index
+        FROM position_by_hash JOIN first_hash
+        ON position_by_hash.height > first_hash.height
+        OR (position_by_hash.height = first_hash.height AND
+            position_by_hash.index >= first_hash.index)
+      ),
+      total_size AS (
+        SELECT COUNT(*) FROM my_txs
+      ),
+      t AS (
+        SELECT my_txs.height, my_txs.index
+        FROM my_txs
+        JOIN previous_txes ON my_txs.height = previous_txes.height
+        AND my_txs.index = previous_txes.index
+        LIMIT :page_size
+      )
+      SELECT height, index, count, perm FROM t
+      RIGHT OUTER JOIN has_perms ON TRUE
```

Comment:
Quite strange join. Maybe place a todo to rewrite it?

Reply:
We want to add permissions to the result of the query, that's why we join

Reply:
It is strange that we use `RIGHT OUTER JOIN .. ON TRUE`, while it seems we could use just `CROSS JOIN`